File: labs/17-workflow-orchestration/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"
	"time"

	openaiclient "github.com/ZYZ-Labs/agent-labs/shared/config"
)

type taskFunc func(deps map[string]interface{}) (interface{}, error)

type Task struct {
	Name    string
	Func    taskFunc
	Deps    []string
	Retries int
	Result  interface{}
	Error   error
}

type WorkflowEngine struct {
	tasks map[string]*Task
}

func newWorkflowEngine() *WorkflowEngine {
	return &WorkflowEngine{tasks: map[string]*Task{}}
}

func (e *WorkflowEngine) addTask(t *Task) {
	e.tasks[t.Name] = t
}

func (e *WorkflowEngine) topologicalSort() ([]string, error) {
	inDegree := map[string]int{}
	dependents := map[string][]string{}
	for name := range e.tasks {
		inDegree[name] = 0
		dependents[name] = nil
	}
	for name, t := range e.tasks {
		for _, dep := range t.Deps {
			if _, ok := e.tasks[dep]; !ok {
				return nil, fmt.Errorf("task %s depends on unknown task %s", name, dep)
			}
			inDegree[name]++
			dependents[dep] = append(dependents[dep], name)
		}
	}

	var queue []string
	for name, deg := range inDegree {
		if deg == 0 {
			queue = append(queue, name)
		}
	}

	var ordered []string
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		ordered = append(ordered, current)
		for _, dependent := range dependents[current] {
			inDegree[dependent]--
			if inDegree[dependent] == 0 {
				queue = append(queue, dependent)
			}
		}
	}

	if len(ordered) != len(e.tasks) {
		return nil, fmt.Errorf("cycle detected in task dependencies")
	}
	return ordered, nil
}

func (e *WorkflowEngine) run() (map[string]interface{}, error) {
	order, err := e.topologicalSort()
	if err != nil {
		return nil, err
	}
	log.Printf("Execution order: %v", order)
	for _, name := range order {
		t := e.tasks[name]
		depsResults := map[string]interface{}{}
		for _, dep := range t.Deps {
			depsResults[dep] = e.tasks[dep].Result
		}
		for attempt := 1; attempt <= t.Retries; attempt++ {
			log.Printf("Running task '%s' (attempt %d/%d)", name, attempt, t.Retries)
			result, err := t.Func(depsResults)
			if err == nil {
				t.Result = result
				t.Error = nil
				break
			}
			t.Error = err
			log.Printf("Task '%s' attempt %d failed: %v", name, attempt, err)
			if attempt == t.Retries {
				log.Printf("Task '%s' exhausted retries", name)
				return nil, err
			}
			time.Sleep(time.Duration(attempt) * 500 * time.Millisecond)
		}
		log.Printf("Task '%s' completed", name)
	}

	results := map[string]interface{}{}
	for name, t := range e.tasks {
		results[name] = t.Result
	}
	return results, nil
}

func fetchData(deps map[string]interface{}) (interface{}, error) {
	return map[string]string{
		"title":   "AI Agent Engineering",
		"content": "Workflow orchestration is essential for reliable agent systems.",
	}, nil
}

func makeAnalyzeSentiment(client *openaiclient.Client) taskFunc {
	return func(deps map[string]interface{}) (interface{}, error) {
		fetch := deps["fetch"].(map[string]string)
		text := fetch["content"]
		if client == nil {
			log.Println("No LLM available; using deterministic sentiment fallback")
			return "positive", nil
		}
		resp, err := client.ChatCompletion(openaiclient.ChatCompletionRequest{
			Messages: []openaiclient.Message{
				{Role: "system", Content: "Classify sentiment as exactly one word: positive, negative, or neutral."},
				{Role: "user", Content: text},
			},
			Temperature: floatPtr(0.0),
			MaxTokens:   intPtr(10),
		})
		if err != nil {
			return "", err
		}
		return strings.ToLower(strings.TrimSpace(resp.Choices[0].Message.Content)), nil
	}
}

func makeFlakyQualityCheck() taskFunc {
	callCount := 0
	return func(deps map[string]interface{}) (interface{}, error) {
		callCount++
		if callCount < 3 {
			return "", fmt.Errorf("quality check service unavailable (attempt %d)", callCount)
		}
		sentiment := deps["sentiment"].(string)
		return fmt.Sprintf("quality_ok (%s)", sentiment), nil
	}
}

func makeSummarize(client *openaiclient.Client) taskFunc {
	return func(deps map[string]interface{}) (interface{}, error) {
		fetch := deps["fetch"].(map[string]string)
		sentiment := deps["sentiment"].(string)
		if client == nil {
			log.Println("No LLM available; using deterministic summary fallback")
			return fmt.Sprintf("Summary: '%s' has %s sentiment.", fetch["title"], sentiment), nil
		}
		prompt := fmt.Sprintf("Title: %s\nContent: %s\nSentiment: %s\nWrite a one-sentence summary.", fetch["title"], fetch["content"], sentiment)
		resp, err := client.ChatCompletion(openaiclient.ChatCompletionRequest{
			Messages:    []openaiclient.Message{{Role: "user", Content: prompt}},
			Temperature: floatPtr(0.0),
			MaxTokens:   intPtr(60),
		})
		if err != nil {
			return "", err
		}
		return strings.TrimSpace(resp.Choices[0].Message.Content), nil
	}
}

func floatPtr(f float64) *float64 {
	return &f
}

func intPtr(i int) *int {
	return &i
}

func main() {
	cfg := openaiclient.LoadConfig()
	var client *openaiclient.Client
	if cfg.APIKey != "" || strings.HasPrefix(cfg.BaseURL, "http://localhost") {
		client = openaiclient.NewClient()
	} else {
		log.Println("LLM client disabled: OPENAI_API_KEY not set")
	}

	engine := newWorkflowEngine()
	engine.addTask(&Task{Name: "fetch", Func: fetchData, Retries: 2})
	engine.addTask(&Task{Name: "sentiment", Func: makeAnalyzeSentiment(client), Deps: []string{"fetch"}, Retries: 2})
	engine.addTask(&Task{Name: "quality", Func: makeFlakyQualityCheck(), Deps: []string{"sentiment"}, Retries: 5})
	engine.addTask(&Task{Name: "summary", Func: makeSummarize(client), Deps: []string{"fetch", "sentiment"}, Retries: 2})

	fmt.Println("Starting DAG workflow...")
	results, err := engine.run()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Workflow failed: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("\nFinal results:")
	b, _ := json.MarshalIndent(results, "", "  ")
	fmt.Println(string(b))
}

File: labs/17-workflow-orchestration/java/Main.java
import com.agentlabs.OpenAiClient;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.*;

public class Main {

    public static void main(String[] args) throws Exception {
        OpenAiClient client = null;
        try {
            client = new OpenAiClient();
        } catch (IllegalStateException exc) {
            System.out.println("LLM client disabled: " + exc.getMessage());
        }

        WorkflowEngine engine = new WorkflowEngine();
        engine.addTask(new Task("fetch", ctx -> fetchData()));
        engine.addTask(new Task("sentiment", ctx -> analyzeSentiment(client, ctx), List.of("fetch")));
        engine.addTask(new Task("quality", ctx -> flakyQualityCheck(ctx), List.of("sentiment"), 5));
        engine.addTask(new Task("summary", ctx -> summarize(client, ctx), List.of("fetch", "sentiment")));

        System.out.println("Starting DAG workflow...");
        Map<String, Object> results = engine.run();
        System.out.println("\nFinal results:");
        System.out.println(new ObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(results));
    }

    private static Map<String, String> fetchData() {
        return Map.of(
                "title", "AI Agent Engineering",
                "content", "Workflow orchestration is essential for reliable agent systems."
        );
    }

    private static String analyzeSentiment(OpenAiClient client, Map<String, Object> ctx) throws Exception {
        @SuppressWarnings("unchecked")
        Map<String, String> fetch = (Map<String, String>) ctx.get("fetch");
        String text = fetch.get("content");
        if (client == null) {
            System.out.println("No LLM available; using deterministic sentiment fallback");
            return "positive";
        }
        ObjectMapper mapper = new ObjectMapper();
        ObjectNode payload = mapper.createObjectNode();
        ArrayNode messages = mapper.createArrayNode();
        ObjectNode system = messages.addObject();
        system.put("role", "system");
        system.put("content", "Classify sentiment as exactly one word: positive, negative, or neutral.");
        ObjectNode user = messages.addObject();
        user.put("role", "user");
        user.put("content", text);
        payload.set("messages", messages);
        payload.put("temperature", 0.0);
        payload.put("max_tokens", 10);

        JsonNode response = client.chatCompletion(payload);
        return client.extractMessage(response).get("content").asText().strip().toLowerCase();
    }

    private static FlakyChecker flakyChecker = new FlakyChecker();

    private static String flakyQualityCheck(Map<String, Object> ctx) {
        String sentiment = (String) ctx.get("sentiment");
        return flakyChecker.check(sentiment);
    }

    private static String summarize(OpenAiClient client, Map<String, Object> ctx) throws Exception {
        @SuppressWarnings("unchecked")
        Map<String, String> fetch = (Map<String, String>) ctx.get("fetch");
        String sentiment = (String) ctx.get("sentiment");
        if (client == null) {
            System.out.println("No LLM available; using deterministic summary fallback");
            return "Summary: '" + fetch.get("title") + "' has " + sentiment + " sentiment.";
        }
        String prompt = "Title: " + fetch.get("title") + "\n"
                + "Content: " + fetch.get("content") + "\n"
                + "Sentiment: " + sentiment + "\n"
                + "Write a one-sentence summary.";
        ObjectMapper mapper = new ObjectMapper();
        ObjectNode payload = mapper.createObjectNode();
        ArrayNode messages = mapper.createArrayNode();
        ObjectNode user = messages.addObject();
        user.put("role", "user");
        user.put("content", prompt);
        payload.set("messages", messages);
        payload.put("temperature", 0.0);
        payload.put("max_tokens", 60);

        JsonNode response = client.chatCompletion(payload);
        return client.extractMessage(response).get("content").asText().strip();
    }

    @FunctionalInterface
    interface TaskFunc {
        Object run(Map<String, Object> ctx) throws Exception;
    }

    static class Task {
        final String name;
        final TaskFunc func;
        final List<String> deps;
        final int retries;
        Object result;
        Exception error;

        Task(String name, TaskFunc func) {
            this(name, func, List.of(), 2);
        }

        Task(String name, TaskFunc func, List<String> deps) {
            this(name, func, deps, 2);
        }

        Task(String name, TaskFunc func, List<String> deps, int retries) {
            this.name = name;
            this.func = func;
            this.deps = deps != null ? deps : List.of();
            this.retries = retries;
        }
    }

    static class WorkflowEngine {
        private final Map<String, Task> tasks = new LinkedHashMap<>();

        void addTask(Task task) {
            tasks.put(task.name, task);
        }

        Map<String, Object> run() throws Exception {
            List<String> order = topologicalSort();
            System.out.println("Execution order: " + order);
            for (String name : order) {
                Task task = tasks.get(name);
                Map<String, Object> depResults = new HashMap<>();
                for (String dep : task.deps) {
                    depResults.put(dep, tasks.get(dep).result);
                }
                for (int attempt = 1; attempt <= task.retries; attempt++) {
                    try {
                        System.out.println("Running task '" + task.name + "' (attempt " + attempt + "/" + task.retries + ")");
                        task.result = task.func.run(depResults);
                        task.error = null;
                        break;
                    } catch (Exception exc) {
                        task.error = exc;
                        System.out.println("Task '" + task.name + "' attempt " + attempt + " failed: " + exc.getMessage());
                        if (attempt == task.retries) {
                            System.err.println("Task '" + task.name + "' exhausted retries");
                            throw exc;
                        }
                        Thread.sleep(500L * attempt);
                    }
                }
                System.out.println("Task '" + task.name + "' completed");
            }
            Map<String, Object> results = new LinkedHashMap<>();
            for (Task task : tasks.values()) {
                results.put(task.name, task.result);
            }
            return results;
        }

        private List<String> topologicalSort() {
            Map<String, Integer> inDegree = new HashMap<>();
            Map<String, List<String>> dependents = new HashMap<>();
            for (String name : tasks.keySet()) {
                inDegree.put(name, 0);
                dependents.put(name, new ArrayList<>());
            }
            for (Task task : tasks.values()) {
                for (String dep : task.deps) {
                    if (!tasks.containsKey(dep)) {
                        throw new IllegalArgumentException("Task " + task.name + " depends on unknown task " + dep);
                    }
                    inDegree.put(task.name, inDegree.get(task.name) + 1);
                    dependents.get(dep).add(task.name);
                }
            }
            List<String> queue = new ArrayList<>();
            for (Map.Entry<String, Integer> entry : inDegree.entrySet()) {
                if (entry.getValue() == 0) {
                    queue.add(entry.getKey());
                }
            }
            List<String> ordered = new ArrayList<>();
            while (!queue.isEmpty()) {
                String current = queue.remove(0);
                ordered.add(current);
                for (String dependent : dependents.get(current)) {
                    inDegree.put(dependent, inDegree.get(dependent) - 1);
                    if (inDegree.get(dependent) == 0) {
                        queue.add(dependent);
                    }
                }
            }
            if (ordered.size() != tasks.size()) {
                throw new IllegalArgumentException("Cycle detected in task dependencies");
            }
            return ordered;
        }
    }

    static class FlakyChecker {
        private int callCount = 0;

        String check(String sentiment) {
            callCount++;
            if (callCount < 3) {
                throw new RuntimeException("Quality check service unavailable (attempt " + callCount + ")");
            }
            return "quality_ok (" + sentiment + ")";
        }
    }
}
